ssz/src/bitfield/bitvector_dynamic.rs

```diff
@@ -0,0 +1,512 @@
+//! Provides `Bitfield<Dynamic>` (BitVectorDynamic)
+/// for encoding and decoding bitvectors that have a dynamic length.
+use crate::{
+    bitfield::{bytes_for_bit_len, Bitfield, BitfieldBehaviour, Error, SMALLVEC_LEN},
+    Decode, DecodeError, Encode,
+};
+use core::marker::PhantomData;
+use serde::de::{Deserialize, Deserializer};
+use serde::ser::{Serialize, Serializer};
+use serde_utils::hex::{encode as hex_encode, PrefixedHexVisitor};
+use smallvec::{smallvec, SmallVec, ToSmallVec};
+
+/// A marker struct used to declare dynamic length behaviour on a Bitfield.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Dynamic;
+
+/// A heap-allocated, ordered collection of `bool` values with a length set at runtime.
+pub type BitVectorDynamic = Bitfield<Dynamic>;
+
+impl BitfieldBehaviour for Dynamic {}
+
+impl Bitfield<Dynamic> {
+    /// Create a new dynamic bitfield with the given length (all bits false).
+    /// Length must be a multiple of 8 bits and greater than 0.
+    pub fn new(len: usize) -> Result<Self, Error> {
+        if len == 0 {
+            return Err(Error::InvalidByteCount {
+                given: len,
+                expected: 8,
+            });
+        }
+        if len % 8 != 0 {
+            return Err(Error::InvalidByteCount {
+                given: len,
+                expected: (len / 8 + 1) * 8,
+            });
+        }
+        Ok(Self {
+            bytes: smallvec![0; bytes_for_bit_len(len)],
+            len,
+            _phantom: PhantomData,
+        })
+    }
+
+    pub fn into_bytes(self) -> SmallVec<[u8; SMALLVEC_LEN]> {
+        self.into_raw_bytes()
+    }
+
+    /// Create a dynamic bitfield from raw bytes and a declared logical length.
+    /// Can be used to specify a max_length if called directly instead of via Decode.
+    pub fn from_bytes_with_len(
+        bytes: SmallVec<[u8; SMALLVEC_LEN]>,
+        len: usize,
+    ) -> Result<Self, Error> {
+        if len != bytes.len() * 8 {
+            return Err(Error::InvalidByteCount {
+                given: len,
+                expected: bytes.len() * 8,
+            });
+        }
+        Self::from_raw_bytes(bytes, len)
+    }
+
+    /// Compute the intersection of two bitfields.
+    pub fn intersection(&self, other: &Self) -> Result<Self, Error> {
+        let max_len = std::cmp::max(self.len(), other.len());
+        let mut result = Self::new(max_len)?;
+
+        for i in 0..result.bytes.len() {
+            result.bytes[i] =
+                self.bytes.get(i).copied().unwrap_or(0) & other.bytes.get(i).copied().unwrap_or(0);
+        }
```

Comment:
We could remove the `[i]` array access here with:

```rust
for (i, byte) in result.bytes.iter_mut().enumerate() {
    *byte = ...
}
```

Although `[i]` isn't going to panic here, the `iter_mut` proves that to the compiler.